### DISCUSSION TITLE
RavenDB Changes API with Reactive extensions (Rx.Net) freezes the app if I dispose the subscription.

### DISCUSSION THREAD

**Discussion by Mike-Logit** [ravendb/ravendb#13958] [Q&A]
I noticed that if I dispose the subscription, [like the documentation suggests](https://ravendb.org/docs/article-page/5.2/csharp/client-api/changes/what-is-changes-api#unsubscribing), then the application freezes when I use the Changes API with Rx.Net.

It seems to be this line in the source code where `_task.Wait()` will wait forever and freeze the app:
https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Client/Documents/Changes/DatabaseChanges.cs#L367

I went with Rx.Net because I wanted the timeout logic. My Asp.Net 5 app receives a GET request using AJAX  and I wait for 5-8 seconds to see if a document has updated with a predicate (using FirstOrDefault). It works great if I don't dispose of the subscription but I'm worried about leaking resources each time I do one of these GET requests.

This is what it looks like when using dispose:

```csharp
public async Task<TDto> SubscribeToCreateOrUpdateAsync(Predicate<TDto> predicate, int timeoutInSeconds)
{
    var subscription = DocumentStore.Changes();

    try
    {
        await subscription.EnsureConnectedNow();
        var observable = subscription.ForDocumentsInCollection<TDocument>();
        await observable.EnsureSubscribedNow();

        var result = await observable
            .Timeout(TimeSpan.FromSeconds(timeoutInSeconds))
            .Catch<DocumentChange, TimeoutException>(_ => Observable.Return(new DocumentChange()))
            .Where(changes => !string.IsNullOrWhiteSpace(changes.Id) && changes.Type is DocumentChangeTypes.Put)
            .Select(async change => await LoadAsync(change.Id))
            .Select(t => t.Result)
            .FirstOrDefaultAsync(dto => predicate(dto))
            .ToTask();

        return result;
    }
    finally
    {
        // This freezes the application (internally, it uses _task.Wait() which never exists:
        subscription?.Dispose();
    }
}
```

However, this code works perfectly:

```csharp
public async Task<TDto> SubscribeToCreateOrUpdateAsync(Predicate<TDto> predicate, int timeoutInSeconds)
{
    var subscription = DocumentStore.Changes();

    await subscription.EnsureConnectedNow();
    var observable = subscription.ForDocumentsInCollection<TDocument>();
    await observable.EnsureSubscribedNow();

    var result = await observable
        .Timeout(TimeSpan.FromSeconds(timeoutInSeconds))
        .Catch<DocumentChange, TimeoutException>(_ => Observable.Return(new DocumentChange()))
        .Where(changes => !string.IsNullOrWhiteSpace(changes.Id) && changes.Type is DocumentChangeTypes.Put)
        .Select(async change => await LoadAsync(change.Id))
        .Select(t => t.Result)
        .FirstOrDefaultAsync(dto => predicate(dto))
        .ToTask();

    return result;
}
```

Is this safe? Thanks for any feedback.

EDIT: I'm using version 4.2.8 of RavenDB but the issue still seems to occur when testing with the latest version.

**Comment by ayende** [comment]
You are _expected_ to to call the `Dispose` on the `observable`, not the result of `Changes()`.

My guess is that you are currently busy inside Rx, which isn't letting us release the Changes connection.

**Reply by ayende** [reply]
Double checked the docs, they are wrong.

**Reply by ayende** [reply]
https://issues.hibernatingrhinos.com/issue/RDoc-2151

**Reply by Mike-Logit** [reply]
> You are expected to to call the Dispose on the observable, 

@ayende Thank you for the fast reply. There doesn't seem to be a `Dispose` method on the returned `IChangesObserable` interface returned from the `ForDocumentsInCollection` call, but if I use `.Subscribe` on the returned `IChangesObserable` then I can dispose the `IDisposable` that gets returned.

However, I don't need to use the `.Subscribe` method like in the documentation when using Reactive extensions (as shown above in the original code snippet). So my best guess is that I have nothing to dispose of if I don't call `.Subscribe`?

**Reply by Mike-Logit** [reply]
For example, you'll notice that I never actually call `.Subscribe` in my code, such as in the recommended example:

```
var subscription = changes.ForAllDocuments()
    .Subscribe(change => Console.WriteLine("{0} on document {1}", change.Type, change.Id));
```

I want to use the Reactive extension method `.Timeout` and a predicate with `.FirstOrDefaultAsync`, and `.Catch` to catch the failed time out exception. I have no idea how to do this with only `IChangesObservable.Subscribe`.

**Reply by ayende** [reply]
The problem is more about Rx than RavenDB. You need to compose this properly, might want to look here:

https://stackoverflow.com/questions/39860166/rx-net-perform-asynchronous-side-effect-when-subscription-is-disposed

https://stackoverflow.com/questions/20075543/using-rx-timeout-with-fromeventpattern-observer

Basically, `Timeout.Subscribe` is what you probably need to dispose. When you do that as a chain of calls, you don't have the right seam to call
